In [125]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.preprocessing import StandardScaler


# Part 1: Load the dataset

In [126]:
# Load the given datasets
categorical_df = pd.read_csv("chronic_kidney_disease_categorical.csv")
numerical_df = pd.read_csv("chronic_kidney_disease_numerical.csv")

# Print the data
print("categorical_df")
print(categorical_df)
print("numerical_df")
print(numerical_df)


categorical_df
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good 

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)


Answer: The categorical_df contains categorical features which represents health-related categories and symptoms. The numerical_df contains numerical features which represent different health metrics. The features are used to predict the target outcome. These features are in both the categorical and numerical data frames. The column Target in categorical_df is the label which is the indicator of chronic kidney disease. 3 examples of categorical features are 'al', 'su', 'rbc'. 3 examples of numerical features are 'age', 'bp', and 'bgr'. 

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [127]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows

numerical_duplicates = numerical_df.duplicated(subset=['unique_id']).sum()
print(f"Total number of duplicate rows in numerical_df: {numerical_duplicates}")
numerical_df = numerical_df.drop_duplicates(subset=['unique_id'])

# Drop these duplicate rows


# Repeat the same for categorical dataset. Print the duplicate rows and drop them

categorical_duplicates = categorical_df.duplicated(subset=['unique_id']).sum()
print(f"Total number of duplicate rows in categorical_df: {categorical_duplicates}")
categorical_df = categorical_df.drop_duplicates(subset=['unique_id'])


Total number of duplicate rows in numerical_df: 30
Total number of duplicate rows in categorical_df: 30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

In [128]:
# Merge the two given numerical and categorical datasets based on their unique_ID.

#Print the combined dataset

combined_df = pd.merge(categorical_df, numerical_df, on='unique_id')

print(combined_df)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
385     789906  0.0  0.0    normal    normal  notpresent  notpresent   no   
386     490253  0.0  0.0    normal    normal  notpresent  notpresent   no   
387     492993  3.0  0.0       NaN    normal  notpresent  notpresent  yes   
388     273038  0.0  0.0    normal    normal  notpresent  notpresent   no   
389     583971  1.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm cad  ...    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc 

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [129]:
# Calculate the percentage of rows that contain atleast one missing value

missing_rows_percentage = (combined_df.isnull().any(axis=1).sum() / len(combined_df)) * 100

# Print %
print(f"Percentage of rows with at least one missing value: {missing_rows_percentage:.2f}%")

# Drop these rows from the dataset
combined_df_cleaned = combined_df.dropna()

# Print the Dataset
print(combined_df_cleaned)

Percentage of rows with at least one missing value: 60.77%
     unique_id   al   su       rbc      pc         pcc          ba  htn   dm  \
4       881763  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
5       114717  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
7       901528  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
8       433613  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
9       621332  1.0  0.0  abnormal  normal  notpresent  notpresent   no  yes   
..         ...  ...  ...       ...     ...         ...         ...  ...  ...   
382     996076  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
384     397388  2.0  2.0    normal  normal  notpresent     present  yes   no   
385     789906  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
386     490253  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
388     273038  0.0  0.0    normal  normal  notpresent  notpr

## Part 3.4: Sort the dataset according to the Labels

In [130]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting

# Print the dataset

combined_df_sorted = combined_df_cleaned.sort_values(by='Target').reset_index(drop=True)

print(combined_df_sorted)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       534245  2.0  0.0    normal  abnormal     present  notpresent   no   
1       349892  1.0  3.0  abnormal  abnormal  notpresent  notpresent  yes   
2       343710  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   
4       514721  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
..         ...  ...  ...       ...       ...         ...         ...  ...   
148     489160  0.0  0.0    normal    normal  notpresent  notpresent   no   
149     530138  0.0  0.0    normal    normal  notpresent  notpresent   no   
150     261137  0.0  0.0    normal    normal  notpresent  notpresent   no   
151     322882  0.0  0.0    normal    normal  notpresent  notpresent   no   
152     273038  0.0  0.0    normal    normal  notpresent  notpresent   no   

      dm  cad  ...    bp    bgr     bu    sc    sod  pot  hemo   pcv     wb

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [131]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

# Print the dataset

combined_df_encoded = pd.get_dummies(combined_df_sorted, columns=categorical_columns)
print(combined_df_encoded)

     unique_id   al   su   age    bp    bgr     bu    sc    sod  pot  ...  \
0       534245  2.0  0.0  45.0  70.0  117.0   52.0   2.2  136.0  3.8  ...   
1       349892  1.0  3.0  59.0  70.0  424.0   55.0   1.7  138.0  4.5  ...   
2       343710  2.0  0.0  61.0  80.0  173.0  148.0   3.9  135.0  5.2  ...   
3       102677  3.0  0.0  59.0  70.0   76.0  186.0  15.0  135.0  7.6  ...   
4       514721  2.0  0.0  56.0  90.0  129.0  107.0   6.7  131.0  4.8  ...   
..         ...  ...  ...   ...   ...    ...    ...   ...    ...  ...  ...   
148     489160  0.0  0.0  28.0  60.0   79.0   50.0   0.5  145.0  5.0  ...   
149     530138  0.0  0.0  23.0  80.0  111.0   34.0   1.1  145.0  4.0  ...   
150     261137  0.0  0.0  35.0  60.0  105.0   39.0   0.5  135.0  3.9  ...   
151     322882  0.0  0.0  33.0  80.0  128.0   38.0   0.6  135.0  3.9  ...   
152     273038  0.0  0.0  47.0  60.0  117.0   22.0   1.2  138.0  3.5  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_ye

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [132]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

# Print the dataset


for col in numerical_columns:
    mean = combined_df_encoded[col].mean()
    std_dev = combined_df_encoded[col].std()
    upper_bound = mean + 3 * std_dev
    lower_bound = mean - 3 * std_dev
    combined_df_encoded = combined_df_encoded[(combined_df_encoded[col] >= lower_bound) & (combined_df_encoded[col] <= upper_bound)]

print(combined_df_encoded)

     unique_id   al   su   age    bp    bgr     bu   sc    sod  pot  ...  \
2       343710  2.0  0.0  61.0  80.0  173.0  148.0  3.9  135.0  5.2  ...   
4       514721  2.0  0.0  56.0  90.0  129.0  107.0  6.7  131.0  4.8  ...   
7       124304  3.0  4.0  69.0  70.0  264.0   87.0  2.7  130.0  4.0  ...   
8       980291  3.0  0.0  62.0  70.0  122.0   42.0  1.7  136.0  4.7  ...   
10      615697  4.0  0.0   6.0  60.0   94.0   67.0  1.0  135.0  4.9  ...   
..         ...  ...  ...   ...   ...    ...    ...  ...    ...  ...  ...   
148     489160  0.0  0.0  28.0  60.0   79.0   50.0  0.5  145.0  5.0  ...   
149     530138  0.0  0.0  23.0  80.0  111.0   34.0  1.1  145.0  4.0  ...   
150     261137  0.0  0.0  35.0  60.0  105.0   39.0  0.5  135.0  3.9  ...   
151     322882  0.0  0.0  33.0  80.0  128.0   38.0  0.6  135.0  3.9  ...   
152     273038  0.0  0.0  47.0  60.0  117.0   22.0  1.2  138.0  3.5  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  \
2      

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [133]:
# Normalize the all Numerical Attributes in the dataset.

# Print the dataset

scaler = StandardScaler()

combined_df_encoded[numerical_columns] = scaler.fit_transform(combined_df_encoded[numerical_columns])

print(combined_df_encoded)

     unique_id   al   su       age        bp       bgr        bu        sc  \
2       343710  2.0  0.0  0.810190  0.730553  1.146438  4.014305  1.690051   
4       514721  2.0  0.0  0.494747  1.753327  0.170164  2.475789  3.577798   
7       124304  3.0  4.0  1.314898 -0.292221  3.165549  1.725293  0.881016   
8       980291  3.0  0.0  0.873279 -0.292221  0.014848  0.036678  0.206821   
10      615697  4.0  0.0 -2.659681 -1.314996 -0.606418  0.974798 -0.265116   
..         ...  ...  ...       ...       ...       ...       ...       ...   
148     489160  0.0  0.0 -1.271733 -1.314996 -0.939238  0.336877 -0.602214   
149     530138  0.0  0.0 -1.587175  0.730553 -0.229221 -0.263520 -0.197696   
150     261137  0.0  0.0 -0.830113 -1.314996 -0.362349 -0.075896 -0.602214   
151     322882  0.0  0.0 -0.956290  0.730553  0.147976 -0.113421 -0.534794   
152     273038  0.0  0.0 -0.073050 -1.314996 -0.096093 -0.713817 -0.130277   

          sod       pot  ...  cad_no  cad_yes  appet_good  appe

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The 'unique_id' column in this dataset is not appropriate for modeling and predictions as it is just an identifier and does not provide any useful information that would help to predict the target variable. 

In [134]:
#Remove that column

# Print the dataset

combined_df_final = combined_df_encoded.drop(columns=['unique_id'])

print(combined_df_final)

      al   su       age        bp       bgr        bu        sc       sod  \
2    2.0  0.0  0.810190  0.730553  1.146438  4.014305  1.690051 -0.959702   
4    2.0  0.0  0.494747  1.753327  0.170164  2.475789  3.577798 -1.663925   
7    3.0  4.0  1.314898 -0.292221  3.165549  1.725293  0.881016 -1.839980   
8    3.0  0.0  0.873279 -0.292221  0.014848  0.036678  0.206821 -0.783646   
10   4.0  0.0 -2.659681 -1.314996 -0.606418  0.974798 -0.265116 -0.959702   
..   ...  ...       ...       ...       ...       ...       ...       ...   
148  0.0  0.0 -1.271733 -1.314996 -0.939238  0.336877 -0.602214  0.800855   
149  0.0  0.0 -1.587175  0.730553 -0.229221 -0.263520 -0.197696  0.800855   
150  0.0  0.0 -0.830113 -1.314996 -0.362349 -0.075896 -0.602214 -0.959702   
151  0.0  0.0 -0.956290  0.730553  0.147976 -0.113421 -0.534794 -0.959702   
152  0.0  0.0 -0.073050 -1.314996 -0.096093 -0.713817 -0.130277 -0.431535   

          pot      hemo  ...  cad_no  cad_yes  appet_good  appet_poor  pe_n

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [135]:
# Export the dataframe to a new csv file

# Export the dataframe to a new json file

combined_df_final.to_csv('cleaned_dataset.csv', index=False)

combined_df_final.to_json('cleaned_dataset.json')


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [136]:
print(combined_df_final.head(15))

     al   su       age        bp       bgr        bu        sc       sod  \
2   2.0  0.0  0.810190  0.730553  1.146438  4.014305  1.690051 -0.959702   
4   2.0  0.0  0.494747  1.753327  0.170164  2.475789  3.577798 -1.663925   
7   3.0  4.0  1.314898 -0.292221  3.165549  1.725293  0.881016 -1.839980   
8   3.0  0.0  0.873279 -0.292221  0.014848  0.036678  0.206821 -0.783646   
10  4.0  0.0 -2.659681 -1.314996 -0.606418  0.974798 -0.265116 -0.959702   
11  3.0  1.0  0.431659  0.730553  2.056147  1.199946  1.690051 -0.607591   
13  1.0  0.0 -0.136138 -1.314996  0.924557  1.912917  1.285533  0.096632   
14  4.0  2.0  0.999456  2.776102  0.924557  0.486976  3.914895 -0.079424   
15  4.0  0.0 -0.009961  0.730553 -0.584229  4.577176  4.251993 -0.783646   
16  4.0  3.0  1.314898 -0.292221  2.056147  2.063016  3.308119 -3.600537   
17  3.0  2.0  1.567253  2.776102  3.853379  1.837868  2.836183 -0.079424   
20  4.0  0.0 -1.713353  1.753327 -0.317973 -0.038371  0.206821 -2.720258   
21  4.0  1.0

### Paste the markdown table here

| al   | su   | age       | bp        | bgr       | bu        | sc       | sod       | pot       | hemo       | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|------|------|-----------|-----------|-----------|-----------|----------|-----------|-----------|------------|--------|---------|------------|-------------|-------|--------|--------|---------|-------------|----------------|
| 2.0  | 0.0  | 0.810190  | 0.730553  | 1.146438  | 4.014305  | 1.690051 | -0.959702 | 1.373024  | -2.913221  | 0      | 1       | 0          | 1           | 0     | 1      | 0      | 1       | 1           | 0              |
| 2.0  | 0.0  | 0.494747  | 1.753327  | 0.170164  | 2.475789  | 3.577798 | -1.663925 | 0.743327  | -2.304078  | 1      | 0       | 1          | 0           | 1     | 0      | 1      | 0       | 1           | 0              |
| 3.0  | 4.0  | 1.314898  | -0.292221 | 3.165549  | 1.725293  | 0.881016 | -1.839980 | -0.516068 | -0.824731  | 0      | 1       | 1          | 0           | 0     | 1      | 1      | 0       | 1           | 0              |
| 3.0  | 0.0  | 0.873279  | -0.292221 | 0.014848  | 0.036678  | 0.206821 | -0.783646 | 0.585903  | -0.781221  | 1      | 0       | 1          | 0           | 1     | 0      | 1      | 0       | 1           | 0              |
| 4.0  | 0.0  | -2.659681 | -1.314996 | -0.606418 | 0.974798  | -0.265116| -0.959702 | 0.900751  | -1.955996  | 1      | 0       | 0          | 1           | 1     | 0      | 1      | 0       | 1           | 0              |
| 3.0  | 1.0  | 0.431659  | 0.730553  | 2.056147  | 1.199946  | 1.690051 | -0.607591 | 0.900751  | -1.520894  | 1      | 0       | 1          | 0           | 0     | 1      | 1      | 0       | 1           | 0              |
| 1.0  | 0.0  | -0.136138 | -1.314996 | 0.924557  | 1.912917  | 1.285533 | 0.096632  | -0.516068 | -1.999507  | 1      | 0       | 1          | 0           | 1     | 0      | 1      | 0       | 1           | 0              |
| 4.0  | 2.0  | 0.999456  | 2.776102  | 0.924557  | 0.486976  | 3.914895 | -0.079424 | 0.428478  | -2.826200  | 1      | 0       | 1          | 0           | 0     | 1      | 0      | 1       | 1           | 0              |
| 4.0  | 0.0  | -0.009961 | 0.730553  | -0.584229 | 4.577176  | 4.251993 | -0.783646 | -0.830916 | -1.999507  | 1      | 0       | 1          | 0           | 1     | 0      | 0      | 1       | 1           | 0              |
| 4.0  | 3.0  | 1.314898  | -0.292221 | 2.056147  | 2.063016  | 3.308119 | -3.600537 | -0.673492 | -2.173547  | 0      | 1       | 1          | 0           | 0     | 1      | 0      | 1       | 1           | 0              |
| 3.0  | 2.0  | 1.567253  | 2.776102  | 3.853379  | 1.837868  | 2.836183 | -0.079424 | -2.247736 | -2.260568  | 0      | 1       | 0          | 1           | 1     | 0      | 1      | 0       | 1           | 0              |
| 4.0  | 0.0  | -1.713353 | 1.753327  | -0.317973 | -0.038371 | 0.206821 | -2.720258 | -1.303189 | -2.652160  | 1      | 0       | 1          | 0           | 1     | 0      | 0      | 1       | 1           | 0              |
| 4.0  | 1.0  | 0.999456  | -1.314996 | 2.610848  | 0.637075  | 1.959729 | -0.607591 | 1.687873  | -2.130037  | 1      | 0       | 0          | 1           | 0     | 1      | 1      | 0       | 1           | 0              |
| 4.0  | 1.0  | 0.494747  | -0.292221 | 1.967395  | -0.563718 | 0.206821 | -0.783646 | -0.830916 | 0.741636   | 1      | 0       | 1          | 0           | 1     | 0      | 1      | 0       | 1           | 0              |
| 4.0  | 0.0  | -0.009961 | 0.730553  | 0.258916  | 3.676582  | 4.791349 | -1.487869 | 1.845297  | -1.781956  | 1      | 0       | 1          | 0           | 0     | 1      | 0      | 1       | 1           | 0              |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

To mitigate data conversion hallucinations from LLM conversions, it is important to use strategies that make sure the results are accurate. For example, instead of doing it all at once with a large date set, you could split the data set into smaller ones. You can also provide the model with a lot of information of what needs to be done and in what format. Overall, for the markdown table it generated for me, I think it did a good job as I don't really see any issues. 

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [137]:
# Code to rename all the columns in the dataset

combined_df_final.rename(columns={
    'age': 'Age (years)',
    'bp': 'Blood Pressure (mm/Hg)',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc': 'Red Blood Cells',
    'pc': 'Pus Cell',
    'pcc': 'Pus Cell clumps',
    'ba': 'Bacteria',
    'bgr': 'Blood Glucose Random (mg/dl)',
    'bu': 'Blood Urea (mg/dl)',
    'sc': 'Serum Creatinine (mg/dl)',
    'sod': 'Sodium (mEq/L)',
    'pot': 'Potassium (mEq/L)',
    'hemo': 'Hemoglobin (gms)',
    'pcv': 'Packed Cell Volume',
    'wc': 'White Blood Cell Count (cells/cumm)',
    'rc': 'Red Blood Cell Count (millions/cmm)',
    'htn': 'Hypertension',
    'dm': 'Diabetes Mellitus',
    'cad': 'Coronary Artery Disease',
    'appet': 'Appetite',
    'pe': 'Pedal Edema',
    'ane': 'Anemia',
    'class': 'Class'
}, inplace=True)

print(combined_df_final)

     Albumin  Sugar  Age (years)  Blood Pressure (mm/Hg)  \
2        2.0    0.0     0.810190                0.730553   
4        2.0    0.0     0.494747                1.753327   
7        3.0    4.0     1.314898               -0.292221   
8        3.0    0.0     0.873279               -0.292221   
10       4.0    0.0    -2.659681               -1.314996   
..       ...    ...          ...                     ...   
148      0.0    0.0    -1.271733               -1.314996   
149      0.0    0.0    -1.587175                0.730553   
150      0.0    0.0    -0.830113               -1.314996   
151      0.0    0.0    -0.956290                0.730553   
152      0.0    0.0    -0.073050               -1.314996   

     Blood Glucose Random (mg/dl)  Blood Urea (mg/dl)  \
2                        1.146438            4.014305   
4                        0.170164            2.475789   
7                        3.165549            1.725293   
8                        0.014848            0.0366

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [138]:
# Converted SQL to Pandas code

result = combined_df_final.groupby('Target_ckd').size().reset_index(name='count')
print(result)

   Target_ckd  count
0           0    111
1           1     22


*Describe what the above code does here*

The query counts the number of occurrences of each unique value in the 'Target' column and presents it in a tabular format. It shows how many rows have chronic kidney disease and how many do not have it. In this case, 111 individuals do not and 22 do.